action/protocol/rewarding/protocol.go

```diff
@@ -39,34 +39,50 @@ var (
 	epochRewardHistoryKeyPrefix = []byte("erh")
 	accountKeyPrefix            = []byte("acc")
 	exemptKey                   = []byte("xpt")
+	errInvalidEpoch             = errors.New("invalid start/end epoch number")
 )
 
 // Protocol defines the protocol of the rewarding fund and the rewarding process. It allows the admin to config the
 // reward amount, users to donate tokens to the fund, block producers to grant them block and epoch reward and,
 // beneficiaries to claim the balance into their personal account.
 type Protocol struct {
-	keyPrefix                   []byte
-	addr                        address.Address
-	foundationBonusP2StartEpoch uint64
-	foundationBonusP2EndEpoch   uint64
+	keyPrefix []byte
+	addr      address.Address
+	cfg       genesis.Rewarding
```

Comment:
the P2 bonus extension is not done properly, we just add 2 numbers `P2Start/End` into the `Protocol{}`, business logic to determine bonus partly depends on the `admin{}` stored in the statedb, and partly depends on these 2
see comment in reward.go